pkg/dockerclient/shell.go

```diff
@@ -554,3 +512,64 @@ func (c *ShellClient) getLastNonEmptyLine(lines []string, offset int) string {
 func (c *ShellClient) replaceSingleQuotes(input string) string {
 	return strings.Replace(input, "'", "’", -1)
 }
+
+func (c *ShellClient) resolveDockerBuildNetwork() string {
+
+	// may contain none as a value
+	networkInterface := os.Getenv("NUCLIO_DOCKER_BUILD_NETWORK")
+	if networkInterface == "" {
+		networkInterface = common.GetEnvOrDefaultString("NUCLIO_BUILD_USE_HOST_NET", "host")
+	}
+	switch networkInterface {
+	case "host":
+		fallthrough
+	case "default":
+		fallthrough
+	case "none":
+		return fmt.Sprintf("--network %s", networkInterface)
+	default:
+		return ""
+	}
+}
+
+func (c *ShellClient) build(buildOptions *BuildOptions, buildArgs string, cacheOption string) error {
+	var err error
+	retryOnErrorMessages := []string{
+
+		// when one of the underlying image is gone (from cache)
+		"No such image: sha256:",
+
+		// when overlay image is gone (from disk)
+		"failed to get digest sha256:",
+	}
+
+	runOptions := &cmdrunner.RunOptions{
+		CaptureOutputMode: cmdrunner.CaptureOutputModeStdout,
+		WorkingDir:        &buildOptions.ContextDir,
+	}
+
+	// retry build on predefined errors that occur during race condition and collisions between
+	// shared onbuild layers
+	common.RetryUntilSuccessful(1*time.Hour, 1*time.Minute, func() bool { // nolint: errcheck
+		runResults, runErr := c.runCommand(runOptions,
+			"docker build %s --force-rm -t %s -f %s %s %s .",
+			c.resolveDockerBuildNetwork(),
+			buildOptions.Image,
+			buildOptions.DockerfilePath,
+			cacheOption,
+			buildArgs)
+
+		// preserve error
+		err = runErr
+
+		// retry on a race condition where `--force-rm` removes a cached layer for other function build in process
+		for _, errorMessage := range retryOnErrorMessages {
+			if runErr != nil && strings.HasPrefix(runResults.Stderr, errorMessage) {
+				return false
+			}
+
+		}
+		return true
```

Comment:
Awesome!
I suggest to take this a step further:

- Move this functionality to a common function `common.RetryUntilSuccessfulOnErrorPatterns`. It's a very useful pattern we have it in a lot of places. We know by now that we'll need this in other places
- Use regex matching to support full patterns and not just prefixes so this will be a useful infra for us in the future